Added scripts for ny-cs4all emails

aoby opened the pull request.

ashercodeorg commented on bin/mail/2016-03-29-ny-cs4all/generate-list.rb:
```diff
@@ -0,0 +1,84 @@
+end
+
+# a- Code Studio teachers who have already been through the K-5 PD
+SECTION_IDS = Set.new
```
This is not a constant, please name section_ids.

ashercodeorg commented on bin/mail/2016-03-29-ny-cs4all/generate-list.rb:
```diff
@@ -0,0 +1,84 @@
+k5_pd_teachers = query_dashboard_teachers <<eos
+  SELECT DISTINCT culled_users.id, culled_users.email, culled_users.name
+  FROM (
+    SELECT * FROM users WHERE email like '%@schools.nyc.gov'
```
super nit: If you are capitalzing SQL keywords (which I like to do), s/like/LIKE.

aoby replied:
lol, super nit. Will fix

ashercodeorg commented on bin/mail/2016-03-29-ny-cs4all/generate-list.rb:
```diff
@@ -0,0 +1,84 @@
+  ) AS culled_users
+  JOIN followers ON (followers.student_user_id = culled_users.id)
+  JOIN sections ON (sections.id = followers.section_id)
+  WHERE sections.id in (#{SECTION_IDS.to_a.join(',')})
```
Do we care if this generates an invalid query if SECTION_IDS is empty?

aoby replied:
I don't think so. I don't foresee all the section ids going away. Currently we have thousands. I could add a check in the script and fail if it's empty?

ashercodeorg commented on bin/mail/2016-03-29-ny-cs4all/generate-list.rb:
```diff
@@ -0,0 +1,84 @@
+eos
+puts "#{k5_pd_teachers.length} @schools.nyc.gov k5-pd teachers"
+
+# b- Code Studio teachers who have not been through any PD
```
nit: Please use proper capitalization and punctuation in the comment.

ashercodeorg commented:
There is enough logic there it would be good to see some tests...

ashercodeorg commented on bin/mail/2016-03-29-ny-cs4all/generate-list.rb:
```diff
@@ -0,0 +1,84 @@
+require_relative '../mailing-common/mailing-list-utils'
+require 'set'
+
+# email addresses *@schools.nyc.gov who are (any of):
```
Perhaps "We build a list of email addresses ..."?

ashercodeorg commented on bin/mail/2016-03-29-ny-cs4all/generate-list.rb:
```diff
@@ -0,0 +1,84 @@
+  rows: 10000
+}
+SOLR.query(SOLR_PARAMS).each do |record|
+  next if record['section_id_s'].to_s == ''
```
Prefer record['section_id_s'].to_s.empty?

Actually, shouldn't record['section_id_s'].empty? suffice?

aoby replied:
It can also be null. Absent active_support's blank? method, .to_s == '' is a nice shorthand for nil? or empty?. I'll see if I can import and use blank? or at least add a comment here.

ashercodeorg commented on bin/mail/2016-03-29-ny-cs4all/generate-list.rb:
```diff
@@ -0,0 +1,84 @@
+SOLR.query(SOLR_PARAMS).each do |record|
+  next if record['section_id_s'].to_s == ''
+  SECTION_IDS << record['section_id_s']
+end
```
Is it better to set SECTION_IDs to the SOLR result and eliminate the empty strings via reject!{|s| s.empty?}?

ashercodeorg replied:
Dumb Q: Is there a reason to prefer (not to prefer) querying SOLR over querying the pegasus forms table?

aoby replied:
Because the section_id_s field is embedded in JSON in the data sql field. In solr it's already extracted. Though I could query all forms of that kind in sql and do the section_id extraction in Ruby. What do you think?

ashercodeorg commented on bin/mail/2016-03-29-ny-cs4all/generate-list.rb:
```diff
@@ -0,0 +1,84 @@
+  FROM (
+    SELECT * FROM users WHERE email like '%@schools.nyc.gov'
+  ) AS culled_users
+  JOIN followers ON (followers.student_user_id = culled_users.id)
```
Do you mean followers.student_user_id or followers.user_id?

aoby replied:
student_user_id. This is looking for teachers who attended the workshop. Followers.user_id is the workshop facilitator.

ashercodeorg commented on bin/mail/2016-03-29-ny-cs4all/generate-list.rb:
```diff
@@ -0,0 +1,84 @@
+# remove all the code studio users from hoc_organizers set
+hoc_organizer_code_studio_users.each_key do |email|
+  hoc_organizers.delete(email)
+end
```
Rather than doing the set difference manually, does it make sense to use the Set class?

aoby replied:
hoc_organizer_code_studio_users is a hash, so set won't work, but I'll use active_support's [Hash.except](http://www.rubydoc.info/docs/rails/2.3.8/ActiveSupport/CoreExtensions/Hash/Except) method.

aoby commented:
I incorporated your feedback and cleaned up the query script. Please take another look.
I'll work on getting some basic tests up too later today. I don't want to get too bogged down in that as it's a one-time script and the email is supposed to go out today. 

Fortunately in the next sprint we will be getting out of the business of doing these email scripts by integrating with a proper email marketing system :)

ashercodeorg commented on bin/mail/2016-03-29-ny-cs4all/generate-list.rb:
```diff
@@ -0,0 +1,95 @@
+    ) AS culled_users
+    JOIN followers ON (followers.student_user_id = culled_users.id)
+    JOIN sections ON (sections.id = followers.section_id)
+    WHERE sections.id in (#{section_ids.to_a.join(',')})
```
Presumably the query need not end in semicolon (since it is missing)?

aoby replied:
Correct. [Sequel.fetch](http://sequel.jeremyevans.net/rdoc/classes/Sequel/Database.html#method-i-fetch) takes care of that.

ashercodeorg commented on bin/mail/2016-03-29-ny-cs4all/generate-list.rb:
```diff
@@ -0,0 +1,95 @@
+
+# A. Query Code Studio teachers who have already been through the K-5 PD.
+def query_k5_pd_teachers
+  section_ids = Set.new
```
nit: Is there any advantage to making sections_ids a set rather than an array, noting that you cast to an array at the end? Also note that duplicates in the IN clause are fine.

aoby replied:
In that case, no. And I'm pretty sure there won't be duplicate section ids anyway since these sections are each created for their respective workshops. I'll change to array.

ashercodeorg commented:
LGTM. If no tests are going to be written, I assume you have sanity checked the queries (results) against real (be it live or fake) data?

aoby commented on bin/mail/2016-03-29-ny-cs4all/generate-list.rb:
```diff
@@ -0,0 +1,95 @@
+
+all_results = k5_pd_teachers.merge(non_pd_teachers).merge(hoc_organizers_no_code_studio)
+puts "#{all_results.length} teachers."
+export_contacts_to_csv ALL, 'teachers.csv'
```
Neither of us noticed this copy/paste error :( 

This ALL contains the result of all the solr queries via query_subscribed_contacts, or in this case all HOC organizers. Way more than the intended all_results.

aoby commented:
@ashercodeorg I updated with unit tests and a fix for the ALL send issue I emailed about. I already sent the emails, so I'm going to merge this but please take a look and let me know if you have other feedback.

ashercodeorg commented on bin/mail/2016-03-29-ny-cs4all/generate-list-functions.rb:
```diff
@@ -0,0 +1,87 @@
+  k5_pd_teachers = query_dashboard_teachers %Q(
+    SELECT DISTINCT culled_users.id, culled_users.email, culled_users.name
+    FROM (
+      SELECT * FROM users WHERE user_type = 'teacher' AND email like '%@schools.nyc.gov'
```
For future reference, though [soft] deleted accounts are auto-magically managed by ActiveRecord, [soft] deleted accounts are _not_ auto-magically managed by MySQL. Presumably, this and similar queries should have `deleted_at IS NULL` clause as part of the `WHERE`.

aoby replied:
Noted. Thanks.
